File: modules/21-llmops-eval/py/04_human_review.py
"""
Task 4 — Human review + feedback loop  🟡

What this teaches:
  - Not every output can be auto-graded reliably. Low-confidence outputs
    (LLM-judge score below a threshold) go into a review queue for humans
    to label.
  - Human labels close the loop: once labelled, cases are promoted back into
    the versioned eval set so future automated runs benefit from them.
  - The queue is a JSONL file — easy to open in a spreadsheet or a web UI.

How to run:
  # Step A — write low-confidence outputs to the queue:
  uv run python modules/21-llmops-eval/py/04_human_review.py --write-queue

  # Step B — simulate a human labelling (interactive CLI):
  uv run python modules/21-llmops-eval/py/04_human_review.py --label

  # Step C — merge approved labels back into the eval set:
  uv run python modules/21-llmops-eval/py/04_human_review.py --merge

  # Run A+B+C in sequence (demo mode):
  uv run python modules/21-llmops-eval/py/04_human_review.py --demo
"""

from __future__ import annotations

import json
import sys
import time
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from llm_core import ChatMessage, ChatOptions, get_provider

REPO_ROOT = Path(__file__).resolve().parents[3]
DATA_DIR = REPO_ROOT / "modules" / "21-llmops-eval" / "data"
REVIEW_QUEUE_PATH = DATA_DIR / "review_queue.jsonl"
EVAL_SET_PATH = DATA_DIR / "eval_set_v1.json"

# How confident must the LLM judge be before we skip human review?
CONFIDENCE_THRESHOLD = 0.75   # score (0–1); below this → queue for human


# ---------------------------------------------------------------------------
# Data model
# ---------------------------------------------------------------------------

@dataclass
class QueueItem:
    item_id: str           # unique id for this queue entry
    case_id: str           # the original eval case id
    question: str
    context: str
    system_output: str
    judge_score: float     # 0–1 from LLM-as-judge
    judge_reason: str
    queued_at: str         # ISO timestamp
    human_label: str | None = None     # "correct" | "incorrect" | "partial"
    human_note: str | None = None
    labelled_at: str | None = None
    promoted: bool = False             # True once merged back into eval set


# ---------------------------------------------------------------------------
# Step 1 — LLM runner + judge (thin version)
# ---------------------------------------------------------------------------

def judge_output(question: str, output: str, rubric: str, provider: Any) -> tuple[float, str]:
    """Score `output` with an LLM-as-judge. Return (score_0_1, reason).

    TODO 1a: Build a prompt asking the judge to reply with JSON:
             {"score": <0-10>, "reason": "<string>"}
    TODO 1b: Call provider.chat() with temperature=0.
    TODO 1c: Parse JSON; normalise score to 0–1. Default to 0.0 on parse error.
    Return (normalised_score, reason_string).
    """
    raise NotImplementedError("TODO: implement judge_output")


def run_and_queue(cases: list[dict[str, Any]], provider: Any) -> list[QueueItem]:
    """Run a mini eval and collect low-confidence outputs into the queue.

    TODO 2a: For each case, call provider.chat() with the context-stuffed prompt.
    TODO 2b: Call judge_output().
    TODO 2c: If judge_score < CONFIDENCE_THRESHOLD, create a QueueItem and
             append it.
    TODO 2d: Print progress: case_id, judge_score, queued or skipped.
    Return the list of new QueueItem objects.
    """
    raise NotImplementedError("TODO: implement run_and_queue")


# ---------------------------------------------------------------------------
# Step 2 — Write to the review queue (JSONL)
# ---------------------------------------------------------------------------

def write_queue(items: list[QueueItem]) -> None:
    """Append QueueItems to REVIEW_QUEUE_PATH as JSONL.

    TODO 3a: Create DATA_DIR if needed.
    TODO 3b: Open REVIEW_QUEUE_PATH in append mode.
    TODO 3c: For each item, write json.dumps(asdict(item)) + '\\n'.
    Print how many items were queued.
    """
    raise NotImplementedError("TODO: implement write_queue")


# ---------------------------------------------------------------------------
# Step 3 — Human labelling (interactive CLI)
# ---------------------------------------------------------------------------

def label_queue_interactive() -> None:
    """Present each unlabelled item in the queue and collect a human label.

    TODO 4a: Read REVIEW_QUEUE_PATH line-by-line; parse each JSON line.
    TODO 4b: Skip items where human_label is already set.
    TODO 4c: Display: question, context (truncated), system_output, judge_score, judge_reason.
    TODO 4d: Prompt: "Label [correct/incorrect/partial/skip]: "
    TODO 4e: If not skip, set human_label, human_note, labelled_at; write back.
    TODO 4f: Rewrite the entire JSONL file with updated items.
    Print a summary of how many items were labelled.
    """
    raise NotImplementedError("TODO: implement label_queue_interactive")


# ---------------------------------------------------------------------------
# Step 4 — Merge labelled items back into the eval set
# ---------------------------------------------------------------------------

def merge_labels_into_eval_set() -> None:
    """Promote 'correct' labels from the queue into the versioned eval set.

    This closes the human-in-the-loop feedback cycle: human-validated
    correct outputs become new golden examples in the eval set.

    TODO 5a: Read REVIEW_QUEUE_PATH; collect items with human_label == 'correct'
             and promoted == False.
    TODO 5b: Load EVAL_SET_PATH.
    TODO 5c: For each approved item, add a new case to eval_set['cases'] with:
             id = "hq_<item_id>", question, context,
             reference_answer = system_output (human approved it),
             graders = ["contains", "llm_judge"],
             rubric = "Human-approved output".
    TODO 5d: Bump the patch version (e.g. "1.0.0" → "1.0.1").
    TODO 5e: Write the updated eval set back to EVAL_SET_PATH.
    TODO 5f: Mark promoted=True in the queue; rewrite REVIEW_QUEUE_PATH.
    Print how many new cases were added.
    """
    raise NotImplementedError("TODO: implement merge_labels_into_eval_set")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def load_eval_cases() -> list[dict[str, Any]]:
    with open(EVAL_SET_PATH) as f:
        return json.load(f)["cases"]


def main() -> None:
    import argparse
    parser = argparse.ArgumentParser(description="Human review queue manager")
    parser.add_argument("--write-queue", action="store_true", help="Run eval + write low-confidence items to queue")
    parser.add_argument("--label", action="store_true", help="Interactive labelling session")
    parser.add_argument("--merge", action="store_true", help="Merge approved labels into eval set")
    parser.add_argument("--demo", action="store_true", help="Run all three steps sequentially")
    args = parser.parse_args()

    if args.write_queue or args.demo:
        provider = get_provider()
        cases = load_eval_cases()
        items = run_and_queue(cases, provider)
        write_queue(items)

    if args.label or args.demo:
        label_queue_interactive()

    if args.merge or args.demo:
        merge_labels_into_eval_set()

    if not any([args.write_queue, args.label, args.merge, args.demo]):
        parser.print_help()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: modules/21-llmops-eval/ts/04-human-review.ts
/**
 * Task 4 — Human review + feedback loop  🟡
 *
 * What this teaches:
 *   - Low-confidence outputs (LLM-judge score below threshold) go to a review
 *     queue so humans can label them.
 *   - Human labels close the loop: approved outputs become new golden eval cases.
 *   - The queue is a JSONL file for easy portability.
 *
 * How to run:
 *   pnpm tsx modules/21-llmops-eval/ts/04-human-review.ts --write-queue
 *   pnpm tsx modules/21-llmops-eval/ts/04-human-review.ts --label
 *   pnpm tsx modules/21-llmops-eval/ts/04-human-review.ts --merge
 *   pnpm tsx modules/21-llmops-eval/ts/04-human-review.ts --demo
 */

import {
  readFileSync,
  writeFileSync,
  appendFileSync,
  mkdirSync,
} from "node:fs";
import * as readline from "node:readline";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { getProvider } from "@learn-ai/llm-core";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, "../../../");
const DATA_DIR = resolve(REPO_ROOT, "modules/21-llmops-eval/data");
const REVIEW_QUEUE_PATH = resolve(DATA_DIR, "review_queue.jsonl");
const EVAL_SET_PATH = resolve(DATA_DIR, "eval_set_v1.json");

const CONFIDENCE_THRESHOLD = 0.75;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface QueueItem {
  item_id: string;
  case_id: string;
  question: string;
  context: string;
  system_output: string;
  judge_score: number;
  judge_reason: string;
  queued_at: string;
  human_label: string | null;
  human_note: string | null;
  labelled_at: string | null;
  promoted: boolean;
}

interface EvalCase {
  id: string;
  question: string;
  context: string;
  reference_answer: string;
  graders: string[];
  rubric: string;
}

// ---------------------------------------------------------------------------
// Step 1 — Judge output
// ---------------------------------------------------------------------------

/**
 * Score output with an LLM-as-judge. Return [score_0_1, reason].
 *
 * TODO 1a: Build a judge prompt with question, rubric, output.
 *          Ask for JSON: {"score": <0-10>, "reason": "<string>"}.
 * TODO 1b: provider.chat() with temperature 0.
 * TODO 1c: Parse JSON; normalise score /10. Default 0 on error.
 * Return [normalised_score, reason].
 */
async function judgeOutput(
  question: string,
  output: string,
  rubric: string,
  provider: ReturnType<typeof getProvider>,
): Promise<[number, string]> {
  // TODO: implement judgeOutput
  throw new Error("TODO: implement judgeOutput");
}

/**
 * Run a mini eval and collect low-confidence outputs.
 *
 * TODO 2a: For each case, call provider.chat() with context-stuffed prompt.
 * TODO 2b: judgeOutput().
 * TODO 2c: If judge_score < CONFIDENCE_THRESHOLD, create a QueueItem.
 * Return QueueItem[].
 */
async function runAndQueue(
  cases: EvalCase[],
  provider: ReturnType<typeof getProvider>,
): Promise<QueueItem[]> {
  // TODO: implement runAndQueue
  throw new Error("TODO: implement runAndQueue");
}

// ---------------------------------------------------------------------------
// Step 2 — Write queue
// ---------------------------------------------------------------------------

/**
 * Append QueueItems to REVIEW_QUEUE_PATH as JSONL.
 *
 * TODO 3a: mkdirSync(DATA_DIR, { recursive: true }).
 * TODO 3b: appendFileSync with JSON.stringify(item) + '\n' for each item.
 */
function writeQueue(items: QueueItem[]): void {
  // TODO: implement writeQueue
  throw new Error("TODO: implement writeQueue");
}

// ---------------------------------------------------------------------------
// Step 3 — Interactive labelling
// ---------------------------------------------------------------------------

/**
 * Present each unlabelled item and collect a human label.
 *
 * TODO 4a: Read + parse REVIEW_QUEUE_PATH.
 * TODO 4b: Skip already-labelled items.
 * TODO 4c: Display question, system_output, judge_score.
 * TODO 4d: Prompt: "Label [correct/incorrect/partial/skip]: "
 * TODO 4e: Update and rewrite the JSONL.
 *
 * Hint: use readline.createInterface for stdin.
 */
async function labelQueueInteractive(): Promise<void> {
  // TODO: implement labelQueueInteractive
  throw new Error("TODO: implement labelQueueInteractive");
}

// ---------------------------------------------------------------------------
// Step 4 — Merge labels back into eval set
// ---------------------------------------------------------------------------

/**
 * Promote 'correct' labels into the versioned eval set.
 *
 * TODO 5a: Load REVIEW_QUEUE_PATH; filter human_label=='correct' && !promoted.
 * TODO 5b: Load EVAL_SET_PATH.
 * TODO 5c: Add a new case for each approved item (id='hq_<item_id>').
 * TODO 5d: Bump the patch version.
 * TODO 5e: Write updated eval set.
 * TODO 5f: Mark promoted=true; rewrite queue.
 */
function mergeLabelsIntoEvalSet(): void {
  // TODO: implement mergeLabelsIntoEvalSet
  throw new Error("TODO: implement mergeLabelsIntoEvalSet");
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

function parseArgs(): { writeQueue: boolean; label: boolean; merge: boolean; demo: boolean } {
  const argv = process.argv.slice(2);
  return {
    writeQueue: argv.includes("--write-queue"),
    label: argv.includes("--label"),
    merge: argv.includes("--merge"),
    demo: argv.includes("--demo"),
  };
}

async function main(): Promise<void> {
  const args = parseArgs();

  if (!args.writeQueue && !args.label && !args.merge && !args.demo) {
    console.log("Usage: 04-human-review.ts [--write-queue] [--label] [--merge] [--demo]");
    process.exit(1);
  }

  if (args.writeQueue || args.demo) {
    const provider = getProvider();
    const data = JSON.parse(readFileSync(EVAL_SET_PATH, "utf8"));
    const items = await runAndQueue(data.cases as EvalCase[], provider);
    writeQueue(items);
    console.log(`Queued ${items.length} items for review.`);
  }

  if (args.label || args.demo) {
    await labelQueueInteractive();
  }

  if (args.merge || args.demo) {
    mergeLabelsIntoEvalSet();
  }
}

main().catch((e) => { console.error(e); process.exit(1); });
